### PR TITLE
FIX: Make skillMaxUpgradeCount work in all cases

### LINKED ISSUE #1693: API: skillMaxUpgradeCount returns 1 when SP is too small and current level is too high
#1475 added `skillMaxUpgradeCount` formula API.

When the skill point (SP) is not enough to upgrade the skill, this API should return 0. However, due to floating-point imprecision, when SP is too small and the current level is too high, this API may return 1. For example:
```js
ns.print(ns.formulas.bladeburner.skillMaxUpgradeCount("Hyperdrive", 1e50, 1)); // Print 1
```

This is how we are implementing it:
```js
const result = Math.round((m + delta) / this.costInc);
const costOfResultPlus1 = this.calculateCost(currentLevel, (result + 1) as PositiveInteger);
if (costOfResultPlus1 <= cost) {
  return result + 1;
}
const costOfResult = this.calculateCost(currentLevel, result as PositiveInteger);
if (costOfResult <= cost) {
  return result;
}
return result - 1;
```
`result` is 0, `cost` is 1, and `costOfResultPlus1` is 0, so it returns 1. In this case, the current level is too high (1e50), so increasing it by 1 (`result + 1`) won't do anything (`costOfResultPlus1` = 0) (if the player calls `ns.bladeburner.upgradeSkill` with `count` = 1, we will notify them that it's impossible to do that).

At first glance, I think that this behavior is (somewhat ?) reasonable. However, some players may be confused: "Why does it return 1 when it's obvious that I won't be able to upgrade the skill with that amount of SP?".

@d0sboots What do you think about this problem? It only matters when the player is farming int and they don't know the pitfall of floating-point imprecision (`A(big) + B(small) may be exactly A in some cases`), so I don't think that it's a big deal.

### PR DESCRIPTION
**Pull request by d0sboots**
This bug has been lingering a long time. A proper fix required the ability to find the next greater/smaller floating-point number. Fixes: #1693

I also changed the format on the big comment block away from LaTeX, for better readability. LaTeX is readable-enough for formulas that will be rendered, but comments don't get rendered.

The tests pass. I didn't do any manual testing.

### CODE REVIEW

File: src/utils/NextAfter.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  }
+  if (x === y) {
+    // Subtle -0 case here.
+    return y;
+  }
+  if (x === 0) {
+    // y can't be 0, from above
+    return y > 0 ? Number.MIN_VALUE : -Number.MIN_VALUE;
+  }
+  const f64 = Float64Array.of(x, y);
```
Comment (hexagonrecursion) [inline src/utils/NextAfter.ts]:
```suggestion
  const f64 = Float64Array.of(x);
```

File: src/utils/NextAfter.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  const u32 = new Uint32Array(f64.buffer);
```
Comment (hexagonrecursion) [inline src/utils/NextAfter.ts]: The byte order of Float64Array and Uint32Array depends on the platform. Maybe we should replace both Float64Array and Uint32Array with DataView

Comment (d0sboots) [reply src/utils/NextAfter.ts]: DataView makes the code a bit more awkward, but sure. I'm not changing the tests because of the readability hit, and because there is zero chance we will ever be running them on IBM z/Architecture mainframes XD

**Comment by catloversg** [comment]
This new approach is similar to the old one in that it assumes there is a "window" that always contains the true result, and that the true result is always either at one edge of the window (r0 or r2) or at a specific value within it (r1). The old version assumed that ±1 was sufficient, but we later found counterexamples showing that it was not. This new version attempts to construct a more accurate window, but I also found counterexamples showing that this window is not correct in all cases.

Counterexample 1:
baseCost = 1, costInc = 2.5, mult = 1
currentLevel = 1e18, cost = 320000000000000070000 (`nextafter(3.2e20, 1e99)`)

In your Jest test, you expect the true result to be 128, but it's actually 191.
- r0 = 0, r1 = 128, r2 = 256
- calculateCost(1e18, 128) = 320000000000000000000
- calculateCost(1e18, 129) = 320000000000000000000
- calculateCost(1e18, 191) = 320000000000000000000
- calculateCost(1e18, 192) = 640000000000000100000

191 is at the exact boundary, not 128.

This is acceptable in our case because `1e18 + 128 === 1e18 + 191`. In other words, calling `upgradeSkill` with either value produces the same final skill level. However, that will not be true when the true result falls outside the expected [r0, r2] window.

Counterexample 2:
baseCost = 1, costInc = 1, mult = 1
currentLevel = 2048, cost = 2.049e32

- r0 = 20243517480910204, r1 = 20243517480910210, r2 = 20243517480910212
- calculateCost(2048, 20243517480910212) = 2.0490000000000017e+32
- calculateCost(2048, 20243517480910210) = 2.0490000000000006e+32

Since neither r2 nor r1 is affordable, the implementation assumes that r0 is the correct answer. However, `calculateCost(2048, 20243517480910204)` is 2.0490000000000002e+32, which is also not affordable.
The true result is 20243517480910200 (`calculateCost(2048, 20243517480910200) = 2.0489999999999988e+32`).

Counterexample 3:
baseCost = 1, costInc = 1, mult = 1
currentLevel = 1, cost = 1e308

This approach assumes that a valid window can be constructed from the numerically stable estimate. However, if the estimate itself is NaN, the approach breaks down. In this example, the estimate is NaN, which causes the computed window to become invalid as well.

Based on these examples, I think we should stop pursuing approaches that rely on constructing a window and assuming that the true result lies within it. Such an approach would be fine if we could prove that assumption, but in the last two attempts, we handwaved the proof, and we have repeatedly found counterexamples afterward.

I submitted an alternative implementation that uses binary search to find the true result directly. I also added additional Jest tests. The new function is called `calculateMaxUpgradeCountNew`. You can switch between the existing `calculateMaxUpgradeCount` implementation and the new one for testing.

**Comment by d0sboots** [comment]
Before I make a detailed reply, I am adding a *lot* more tests. I'll reply once that's done.

File: src/utils/NextAfter.ts (hunk 2 above)
Comment (d0sboots) [reply src/utils/NextAfter.ts]: So, funny thing: I talked with my friend today who used to work on Chrome. He said that people there were basically split 50/50 on handling endian issues; some of them did, but a lot of them felt it was a waste of time. In other words, there is basically 0 chance that Chrome *itself* will ever run on a big-endian system. XD (I imagine the same is likely true for Firefox.)

**Comment by d0sboots** [comment]
> This new approach is similar to the old one in that it assumes there is a "window" that always contains the true result, and that the true result is always either at one edge of the window (r0 or r2) or at a specific value within it (r1). The old version assumed that ±1 was sufficient, but we later found counterexamples showing that it was not. This new version attempts to construct a more accurate window, but I also found counterexamples showing that this window is not correct in all cases.

This is true, and you're right that the window approach was flawed. More below.

> Counterexample 1: baseCost = 1, costInc = 2.5, mult = 1 currentLevel = 1e18, cost = 320000000000000070000 (`nextafter(3.2e20, 1e99)`)
> 
> In your Jest test, you expect the true result to be 128, but it's actually 191.
> 
> * r0 = 0, r1 = 128, r2 = 256
> * calculateCost(1e18, 128) = 320000000000000000000
> * calculateCost(1e18, 129) = 320000000000000000000
> * calculateCost(1e18, 191) = 320000000000000000000
> * calculateCost(1e18, 192) = 640000000000000100000
> 
> 191 is at the exact boundary, not 128.
> 
> This is acceptable in our case because `1e18 + 128 === 1e18 + 191`. In other words, calling `upgradeSkill` with either value produces the same final skill level. However, that will not be true when the true result falls outside the expected [r0, r2] window.

This is not a counterexample, this is by design. (Which is why I was testing for this value explicitly.) In a similar way to how non-integers are not valid amounts to pass to `calculateCost` and related functions, amounts that can't be added to level in a representable fashion aren't valid either. It's not an *error* to use them (that would be too punishing to users), but they way we calculate makes it clear that, semantically, they're not valid amounts. Therefore, it is not appropriate to return them from `skillMaxUpgradeCount`.

> Counterexample 2: baseCost = 1, costInc = 1, mult = 1 currentLevel = 2048, cost = 2.049e32
> 
> * r0 = 20243517480910204, r1 = 20243517480910210, r2 = 20243517480910212
> * calculateCost(2048, 20243517480910212) = 2.0490000000000017e+32
> * calculateCost(2048, 20243517480910210) = 2.0490000000000006e+32
> 
> Since neither r2 nor r1 is affordable, the implementation assumes that r0 is the correct answer. However, `calculateCost(2048, 20243517480910204)` is 2.0490000000000002e+32, which is also not affordable. The true result is 20243517480910200 (`calculateCost(2048, 20243517480910200) = 2.0489999999999988e+32`).

This is a true counterexample, and I added it to the tests.

> Counterexample 3: baseCost = 1, costInc = 1, mult = 1 currentLevel = 1, cost = 1e308
> 
> This approach assumes that a valid window can be constructed from the numerically stable estimate. However, if the estimate itself is NaN, the approach breaks down. In this example, the estimate is NaN, which causes the computed window to become invalid as well.

This was a case where the old code failed, but not due to the method being flawed (although it was.) Rather, this was due to insufficient care being taken with edge cases. Your approach also has these types of issues. This was revealed after I added tests of this nature.

> Based on these examples, I think we should stop pursuing approaches that rely on constructing a window and assuming that the true result lies within it. Such an approach would be fine if we could prove that assumption, but in the last two attempts, we handwaved the proof, and we have repeatedly found counterexamples afterward.
> 
> I submitted an alternative implementation that uses binary search to find the true result directly. I also added additional Jest tests. The new function is called `calculateMaxUpgradeCountNew`. You can switch between the existing `calculateMaxUpgradeCount` implementation and the new one for testing.

The first thing I did was to *significantly* beef up the tests, including adding a random-based stress test. The last one, in particular, ended up being invaluable not for uncovering bugs in my eventual algorithm (the algorithm turned out to be fine), but rather bugs in my assumptions that led to the test code itself being flawed. Because of that, I have much higher confidence in the final correctness of the result, now.

I reverted your test changes, because mine ended up overlapping them, and also because you changed the method name under test. IMO the right way to test alternate implementations is not to change it in the tests, but rather to rename the main method, so you're certain to test everything end-to-end.

I changed my implementation from a hardcoded window to a linear search. The correctness of this should be much more straightforward to see, even more than your binary search. The *speed* of it is still based on a handwaved assertion that the approximation ends up very close to the true result, but both intuition and testing support this.

My new tests showed failures (in many places) on my old code, and obviously pass now. They also fail on your binary search method, producing at least one infinite loop that I didn't bother to track down.

**Comment by d0sboots** [comment]
Oh, one other thing I was going to mention but forgot in my previous comment: Getting floating-point library code right is *hard*. Having mistakes in a prior version is not necessarily evidence that the prior approach is intrinsically flawed (although it still could be). If you look through the source of libraries that implement things like `exp`, `sin`, `hypot`, etc., you will see comments and techniques that are very similar or identical to the things I'm doing here - the reasoning involved can be tricky, but it is by no means unprecedented.

File: src/Bladeburner/Skill.ts [hunk 3]
```diff
@@ -72,15 +85,39 @@ export class Skill {
...
+    // Sanity checks for bnMult and skill settings.
+    if (currentNodeMults.BladeburnerSkillCost <= 0 || this.baseCost <= 0 || this.costInc <= 0) {
+      throw new Error(
+        `Invalid bnMult or skill settings. BladeburnerSkillCost: ${currentNodeMults.BladeburnerSkillCost}, baseCost: ${this.baseCost}, costInc: ${this.costInc}`,
+      );
+    }
+    if (Number.isNaN(currentLevel) || Number.isNaN(cost)) {
+      return Number.NaN;
+    }
+    if (cost < currentNodeMults.BladeburnerSkillCost * this.baseCost) {
```
Comment (catloversg) [inline src/Bladeburner/Skill.ts]: Can you explain why we need to check this case beforehand?

Before my commit, this line checked `cost < this.baseCost`, but there are cases where this check causes the function to return the wrong value. I added a test for that.

File: test/jest/Bladeburner/Skill.test.ts [hunk 4]
```diff
@@ -1,18 +1,178 @@
...
-import type { PositiveNumber } from "../../../src/types";
+import { Skill } from "../../../src/Bladeburner/Skill";
+import { nextafter } from "../../../src/utils/NextAfter";
+import type { PositiveInteger, PositiveNumber } from "../../../src/types";
+import { currentNodeMults } from "../../../src/BitNode/BitNodeMultipliers";
 
 const hyperdrive = Skills[BladeburnerSkillName.Hyperdrive];
 
+// Given two Uint32s (presumably randomly generated), return a floating point
+// value for testing. The first part of the tuple is an integer, and the 2nd
```
Comment (catloversg) [inline test/jest/Bladeburner/Skill.test.ts]: Optional (IIUC):
```js
// Given two random Uint32 values, construct a test value represented as
// (integer, multiplier), where the numeric value is integer * multiplier.
// The first part of the tuple is an integer ...
```

File: test/jest/Bladeburner/Skill.test.ts [hunk 5]
```diff
@@ -1,18 +1,178 @@
...
+      expect(result).toBe(Infinity);
+    });
+
+    it("Inf arg both", () => {
+      const result = hyperdrive.calculateCost(Infinity, Infinity as PositiveInteger);
+      expect(result).toBe(Infinity);
+    });
+
+    it("Last possible upgrade", () => {
+      const level = 2 ** 537 * 1.6 - 2 ** 486;
```
Comment (catloversg) [inline test/jest/Bladeburner/Skill.test.ts:78]: Can you write a bit more detail about the goal of this test and what kind of error it catches? These numbers look very specific, so I want to know how and why you came up with them.

File: test/jest/Bladeburner/Skill.test.ts [hunk 6]
```diff
@@ -1,18 +1,178 @@
...
+    it("should return 0 when currentLevel is too high for floating-point precision", () => {
+      // At level 1e18, the smallest gap between numbers is 128. The cost of
+      // 128 levels will be 128 * (1 + 2.5 * 127/2 + 2.5 * 1e18) = 128 * (2.5e18) = 3.2e20.
+      const result = hyperdrive.calculateMaxUpgradeCount(1e18, nextafter(3.2e20, 0) as PositiveNumber);
+      expect(result).toBe(0);
+    });
+
+    it("should return correct count at high currentLevel", () => {
+      const result = hyperdrive.calculateMaxUpgradeCount(1e18, nextafter(3.2e20, 1e99) as PositiveNumber);
+      expect(result).toBe(128);
```
Comment (catloversg) [inline test/jest/Bladeburner/Skill.test.ts:255]: Optional: Copy the comment referencing this PR from above to here.

Expecting the value of 128 is very debatable, IMO. I'll explain why in detail later.

File: test/jest/Bladeburner/Skill.test.ts [hunk 7]
```diff
@@ -1,18 +1,178 @@
...
+        baseCost: 1,
+        costInc: 1,
+        mults: {},
+      });
+      const result = testSkill.calculateMaxUpgradeCount(2048, 2.049e32 as PositiveNumber);
+      expect(result).toBe(20243517480910200);
+    });
+
+    it("should return 0 when currentLevel is too high for floating-point precision", () => {
+      // At level 1e18, the smallest gap between numbers is 128. The cost of
```
Comment (catloversg) [inline test/jest/Bladeburner/Skill.test.ts]: Optional: Explain what "smallest gap" means in this context. I assume you mean it's the representable value of `1000000000000000100 - 1e18` (more about it later).

**Comment by catloversg** [comment]
> This is not a counterexample, this is by design. (Which is why I was testing for this value explicitly.) In a similar way to how non-integers are not valid amounts to pass to calculateCost and related functions, amounts that can't be added to level in a representable fashion aren't valid either. It's not an error to use them (that would be too punishing to users), but they way we calculate makes it clear that, semantically, they're not valid amounts. Therefore, it is not appropriate to return them from skillMaxUpgradeCount.

This argument is not wrong, but I think it needs some clarification. The first time I read it, it sounded contradictory to me. Only after rereading it many times and considering it from different perspectives, I agree with it to some extent (I still don't agree with it entirely, though).

For posterity's sake, I'll write down what I think about your argument here, and why I accept the value of 128 even though I don't think it's the *only* correct value (depending on the perspective). This is basically the clarification that I think the argument needs.

First, these are 3 different perspectives.

```js
const budget = 320000000000000070000;
const count = calculateMaxUpgradeCount(1e18, budget);
upgradeSkill(1e18, count);
```
[1]: What you do is *exactly* what you get.
If count is `foo`, then `skillLevel` must be *exactly* `skillLevel + foo`. For example:
- count = 65 => skillLevel = 1000000000000000065 (impossible)
- count = 100 => skillLevel = 1000000000000000100
- count = 128 => skillLevel = 1000000000000000128 (impossible)
- count = 191 => skillLevel = 1000000000000000191 (impossible)

[2]: I guarantee you cannot find a higher `count` such that `upgradeSkill(1e18, count)` is successful.
If count is `foo`, then there is no `foo + PositiveInteger` such that calling `upgradeSkill(1e18, foo + PositiveInteger)` is successful. For example:
- count = 65 => `upgradeSkill(1e18, 66)` is an error (not affordable) (wrong)
- count = 128 => `upgradeSkill(1e18, 129)` is an error (not affordable) (wrong)
- count = 191 => `upgradeSkill(1e18, 192)` is an error (not affordable) (correct)

[3]: Suppose `foo` is the highest *representable* level that you can upgrade to with `budget`, `count` is the *representable* value of `foo - currentLevel`.
At currentLevel = 1e18 and budget = 320000000000000070000, you cannot upgrade to any *representable* level higher than 1000000000000000100. Therefore, `count` is the *representable* value of `1000000000000000100 - currentLevel`.

The first time I read your argument, I thought you were saying that 191 is not an appropriate value because `1e18 + 191` is clearly not `1000000000000000191` (this is not a representable value). This is where I thought it was contradictory. How can 128 be the correct value? In [1], both 128 and 191 are the "wrong" values, and the only correct value is 100.

Given floating-point inaccuracy, [1] is clearly too "strict", and after reading your implementation, I think I understand what's going on. The new approach focuses on [3] while the binary search focuses on [2]. In other words, we have different perspectives about what the true answer should be.
- With [2], I focus on finding the highest `count` such that calling `upgradeSkill` is successful.
- With [3], you focus on finding the highest *representable* affordable level and then return the *representable* gap between it and the current level.

With [3] in mind, I accept the value of 128. 128 is clearly not the highest `count` such that `upgradeSkill(1e18, count)` is successful (from the perspective of [2]), but it's the *representable* value of `1000000000000000100 - 1e18`, and 1000000000000000100 is the highest *representable* level that you can upgrade to with `budget`.

> Your approach also has these types of issues. This was revealed after I added tests of this nature.

What's the issue? I only see the infinite loop that you mentioned, but it only happens because you pass values that it considers invalid. Specifically, that code assumes `currentLevel` and `cost` are always finite (`// Sanity checks for bnMult and skill settings. Callers should already check currentLevel and cost.`). "Callers" are the functions in `src/NetscriptFunctions/Formulas.ts`. Checking these infinite cases in `calculateMaxUpgradeCount` is simple, so it's not a big deal.

The stress test shows cases in which the binary search returns an unexpected value (e.g., `With parameters {level:1.8449903992772817e+76, cost:6.652694954817565e+151, prevCost:6.652694954817564e+151, amount:1.3899679065382986e+75}`), but that's the same case with the 128 vs 191 case.

> I reverted your test changes, because mine ended up overlapping them, and also because you changed the method name under test. IMO the right way to test alternate implementations is not to change it in the tests, but rather to rename the main method, so you're certain to test everything end-to-end.

Those tests are the "stepping stones" for the discussion, not the final tests for merging. I assumed you would switch the calls to different implementations manually for testing. In other words, they are WIP tests for the discussion phase. I set them to the new implementation to demonstrate what I said.

> They also fail on your binary search method, producing at least one infinite loop that I didn't bother to track down.

I explained the infinite loop above.

**Comment by d0sboots** [comment]
I think perspectives [1] and [3] are actually the same, although the way I'm thinking about it is closer to 3.

> ```js
> const budget = 320000000000000070000;
> const count = calculateMaxUpgradeCount(1e18, budget);
> upgradeSkill(1e18, count);
> ```
> 
> [1]: What you do is _exactly_ what you get. If count is `foo`, then `skillLevel` must be _exactly_ `skillLevel + foo`. For example:
> 
> * count = 65 => skillLevel = 1000000000000000065 (impossible)
> * count = 100 => skillLevel = 1000000000000000100
> * count = 128 => skillLevel = 1000000000000000128 (impossible)
> * count = 191 => skillLevel = 1000000000000000191 (impossible)

1e18 + 100 is *not* 1000000000000000100. (Mathematically it is, obviously, but we're talking floating point here.) It is 1000000000000000128. You can be thrown off on this point because JS will only print as many digits as it needs for a precise representation, which is typically 17, and fill in the rest with zeros.

You typically need to use programs that support arbitrary precision, and use the binary representation of the float, if you want an exact decimal representation.

File: src/Bladeburner/Skill.ts (hunk 3 above)
Comment (d0sboots) [reply src/Bladeburner/Skill.ts]: I was being defensive but ended up being too lax. Checking <= 0 would be fine.

File: test/jest/Bladeburner/Skill.test.ts (hunk 7 above)
Comment (d0sboots) [reply test/jest/Bladeburner/Skill.test.ts]: I'm not sure how to explain this more/better without explaining the entirety of floating point representation. 

The fundamental property of how floating point numbers work is that they have a mantissa and an exponent. Since the value is mantissa * 2^exponent, there is a gap between representable numbers which is a power of 2 based on the size of the exponent.

**Comment by d0sboots** [comment]
Ah, you can actually see this relatively simply directly in JS:
```
> BigInt(1e18)
1000000000000000000n
> BigInt(1e18+100)
1000000000000000128n
```

**Comment by d0sboots** [comment]
> What's the issue? I only see the infinite loop that you mentioned, but it only happens because you pass values that it considers invalid. Specifically, that code assumes currentLevel and cost are always finite (// Sanity checks for bnMult and skill settings. Callers should already check currentLevel and cost.). "Callers" are the functions in src/NetscriptFunctions/Formulas.ts. Checking these infinite cases in calculateMaxUpgradeCount is simple, so it's not a big deal.

One of the issues is an out-of-range problem in `calculateCost` that produced NaN despite valid values. I didn't 100% track down the logic of your binary search, but it seemed unlikely to me that it would have produced the correct result with that unfixed.

File: test/jest/Bladeburner/Skill.test.ts (hunk 6 above)
Comment (d0sboots) [reply test/jest/Bladeburner/Skill.test.ts:255]: Done.

File: test/jest/Bladeburner/Skill.test.ts (hunk 5 above)
Comment (d0sboots) [reply test/jest/Bladeburner/Skill.test.ts:78]: Done

File: test/jest/Bladeburner/Skill.test.ts (hunk 4 above)
Comment (d0sboots) [reply test/jest/Bladeburner/Skill.test.ts]: Done.

**Comment by d0sboots** [comment]
Your defensive check made me realize there was yet another edge case, which I added a test and a fix for.

**Comment by d0sboots** [comment]
My previous fix left a bad taste in my mouth, and going back I was able to find another edge case with it. This fix is more general.

All of these later class of bugs I am 90% sure would not occur with the parameters we use in the game, but they are technically valid parameters, so the function should work with them.
